### DISCUSSION TITLE
Basic support for opening/saving PDFs - useful?

### DISCUSSION THREAD

**Discussion by hi5** [tannerhelland/PhotoDemon#541] [Ideas]
_Low priority ❗_ 

To clarify: Just to see if there is an interest, I just extract the page I need using one of the tools I list below and then edit it in PD or other graphics program.

Photoshop (PS) is able to open PDFs as no doubt some other graphics programs.

PS can do quite a bit with PDFs and I don't advocate turning **PhotoDemon** into a PDF editor, _but_ perhaps it would be useful to add some very basic functionality and treat a PDF as a multi layered image e.g. just a raster layer, no text editing or editing individual image/objects on a page. (So just like a screenshot of a PDF page pasted into PD as an image for example).
This could be useful to quickly edit a flyer or use a page from a PDF as basis for some new image(s).

So opening a PDF would "simply" convert each page into a raster layer.  Or it could even ask which pages you'd like to import e.g. page 1 & 2, skip the rest.
Saving as PDF saves it as an IMAGE pdf, e.g. doesn't convert the image back to text etc. That would simply not fit within the scope of a graphics program. 

I do believe it is possible to get an image from a PDF page using gdiplus, but don't ask me how 😸 

Note that several tools exist to "export" one or more pages from a PDF to a PNG so it isn't too much work to do it manually:
* [pdftopng](https://www.xpdfreader.com/)
* [ImageMagick](https://stackoverflow.com/questions/2869908/convert-pdf-to-png-using-imagemagick)

**Comment by tannerhelland** [comment]
Thank you for the thoughtful post, @hi5 .  For once, I am slightly ahead of the curve.  I've been working on PDF support for the past few weeks in a separate branch:

https://github.com/tannerhelland/PhotoDemon/commits/pdf-support/

Basic import support is working very well.  Today I'm working on adding custom page ranges to the import dialog (like "1, 3, 5-10") and after a few more tweaks and UI additions, PDF import should be ready to merge into nightly builds.  I'm using [Google's open-source pdfium library](https://pdfium.googlesource.com/pdfium/) for all the low-level PDF interactions, and so far integration has gone smoothly.

I'll report back when the feature has officially been merged into nightly builds.  Thank you again for your ongoing feedback and ideas - they're always helpful!

**Reply by hi5** [reply]
Perfect, I had sought in issues/discussion for PDF but didn't see any mention of it so I thought I'd make a post about it. Look forward to trying it out when the time comes.

**Reply by hi5** [reply]
Having fun experimenting with the new PDF import feature 👍

**Reply by tannerhelland** [reply]
Thank you for taking PDF support for a spin, @hi5 .  I very much appreciate your ongoing help.

I hope to get PDF export implemented soon!

**Comment by amiramir2124** [comment]
https://khanephotoshop.com/

photoshop tutorials
and
Photoshop products
